File: noNonsense/03_fast_slow/active_recall.java
// --- FAST & SLOW ACTIVE RECALL ---

class ListNode { int val; ListNode next; }

public class FastSlow {
    // Type A: Cycle Detection
    public boolean hasCycle(ListNode head) {
        ListNode slow = ___, fast = ___;
        while (___ != null && ___.next != null) {
            slow = slow.___;
            fast = fast.___.next;
            if (slow == ___) return true;
        }
        return false;
    }

    // Type B: Find Middle
    public ListNode findMiddle(ListNode head) {
        ListNode slow = head, fast = head;
        while (___ != null && ___.next != null) {
            slow = slow.___;
            fast = fast.___.next;
        }
        return ___;
    }
}

File: noNonsense/03_fast_slow/active_recall.py
# --- FAST & SLOW ACTIVE RECALL ---

# Type A: Cycle Detection
def has_cycle(head):
    slow, fast = ___, ___
    
    while ___ and ___.next:
        slow = slow.___
        fast = fast.___.next
        
        if slow == ___:
            return True
            
    return False

# Type B: Find Middle
def find_middle(head):
    slow, fast = head, head
    
    while ___ and ___.next:
        slow = slow.___
        fast = fast.___.next
        
    return ___ # Slow is at the middle
